pkg/sql/sessionprotectedts/session_protected_ts.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package sessionprotectedts
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/protectedts/ptpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/protectedts/ptreconcile"
+	"github.com/cockroachdb/cockroach/pkg/sql/isql"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/uuid"
+	"github.com/cockroachdb/errors"
+)
+
+// SessionMetaType is the meta type for protected timestamp records associated with sessions.
+const SessionMetaType = "sessions"
+
+// MakeRecord makes a protected timestamp record to protect a timestamp on
+// behalf of a session.
+func MakeRecord(
+	recordID uuid.UUID, sessionID []byte, tsToProtect hlc.Timestamp, target *ptpb.Target,
+) *ptpb.Record {
+	return &ptpb.Record{
+		ID:        recordID.GetBytesMut(),
+		Timestamp: tsToProtect,
+		Mode:      ptpb.PROTECT_AFTER,
+		MetaType:  SessionMetaType,
+		Meta:      sessionID,
+		Target:    target,
+	}
+}
+
+// MakeStatusFunc returns a function which determines whether the session
+// implied with this value of meta should be removed by the reconciler.
+func MakeStatusFunc() ptreconcile.StatusFunc {
+	return func(ctx context.Context, txn isql.Txn, meta []byte) (shouldRemove bool, _ error) {
+		row, err := txn.QueryRowEx(ctx, "check-for-dead-session", txn.KV(),
+			sessiondata.NodeUserSessionDataOverride,
+			`SELECT EXISTS (SELECT 1 FROM crdb_internal.cluster_sessions WHERE session_id = $1 AND status IN ('ACTIVE', 'IDLE'))`, string(meta))
+		if err != nil {
+			return false, err
+		}
+		if row == nil {
+			return false, errors.AssertionFailedf("no row returned when checking for a dead session")
+		}
+		sessionIsAlive := tree.MustBeDBool(row[0])
+		return bool(!sessionIsAlive), nil
```

Comment:
[nit] Very minor, really don't change it if it can otherwise merge. I think it is mildly nicer to have this case up in the assignment.

Reply:
you're right, changed